### PR TITLE
Flink: handle rescale properly and refactor statistics

### PR DESCRIPTION
**Pull request by stevenzwu**
close issue #10441

### CODE REVIEW

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/AggregatedStatistics.java [hunk 1]
```diff
@@ -35,28 +35,28 @@ class AggregatedStatistics implements Serializable {
   private final long checkpointId;
   private final StatisticsType type;
   private final Map<SortKey, Long> keyFrequency;
-  private final SortKey[] rangeBounds;
+  private final SortKey[] keySamples;
```
Comment (stevenzwu) [inline flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/AggregatedStatistics.java:38]: this rename is needed so that `AggregatedStatistics` can be used to store both the complete samples and calculated range bounds.

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/AggregatedStatisticsTracker.java [hunk 2]
```diff
@@ -258,8 +258,7 @@ private AggregatedStatistics completedStatistics(long checkpointId) {
             sketch.getK(),
             sketch.getN(),
             sketch.getNumSamples());
-        return AggregatedStatistics.fromRangeBounds(
-            checkpointId, SketchUtil.rangeBounds(downstreamParallelism, comparator, sketch));
+        return AggregatedStatistics.fromKeySamples(checkpointId, sketch.getSamples());
```
Comment (stevenzwu) [inline flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/AggregatedStatisticsTracker.java:261]: returning the complete samples from tracker. coordinator will store both complete samples and ranged bounds

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsCoordinator.java [hunk 3]
```diff
@@ -71,6 +76,7 @@ class DataStatisticsCoordinator implements OperatorCoordinator {
   private transient boolean started;
   private transient AggregatedStatisticsTracker aggregatedStatisticsTracker;
   private transient AggregatedStatistics completedStatistics;
+  private transient AggregatedStatistics globalStatistics;
```
Comment (stevenzwu) [inline flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsCoordinator.java:79]: for sketch statistics type, `globalStatistics` only stores range bounds (a much smaller array)

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsOperator.java [hunk 4]
```diff
@@ -103,14 +105,41 @@ public void initializeState(StateInitializationContext context) throws Exception
...
       } else {
-        LOG.info(
-            "Operator {} subtask {} restoring global statistics state", operatorName, subtaskIndex);
-        this.globalStatistics = globalStatisticsState.get().iterator().next();
+        AggregatedStatistics restoredStatistics = globalStatisticsState.get().iterator().next();
+        // Range bounds is calculated from reservoir samplings with the determined partition
+        // (downstream parallelism). If downstream parallelism changed due to rescale, the computed
+        // range bounds array is not applicable. Operators should request coordinator to recompute
+        // the range bounds array using the new parallelism.
+        if (restoredStatistics.type() == StatisticsType.Sketch
```
Comment (stevenzwu) [inline flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsOperator.java:121]: this handles the rescale scenario for sketch stats (range bounds)

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsOperator.java [hunk 5]
```diff
@@ -139,14 +168,16 @@ public void handleOperatorEvent(OperatorEvent event) {
         operatorName,
         subtaskIndex,
         statisticsEvent.checkpointId());
-    globalStatistics =
+    this.globalStatistics =
         StatisticsUtil.deserializeAggregatedStatistics(
             statisticsEvent.statisticsBytes(), aggregatedStatisticsSerializer);
     checkStatisticsTypeMigration();
-    output.collect(new StreamRecord<>(StatisticsOrRecord.fromStatistics(globalStatistics)));
```
Comment (stevenzwu) [inline flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsOperator.java:146]: it is a bug previously. we actually don't want to apply the new stats immediately during normal aggregation and propagation phase. switch happens at checkpoint boundary.

`applyImmediately` flag is added in this PR to distinguish the stats requested during rescale case. In this case, immediate application is desired.

Comment (pvary) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsOperator.java:146]: Why don't apply the new statistics immediately?

Comment (pvary) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsOperator.java:146]: Oh... i remember.  Don't want to mess with the ongoing files

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/AggregatedStatistics.java (hunk 1 above)
Comment (pvary) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/AggregatedStatistics.java:38]: Would it be better to have a separate class for global stat, and aggregate stat? It's a bit late here, and it's a bit hard to follow when the keys keySamples are actually range bounds, and when they contain the full data...
Maybe tomorrow it will be easier to follow 😀

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsCoordinator.java [hunk 6]
```diff
@@ -189,28 +198,75 @@ private void handleDataStatisticRequest(int subtask, StatisticsEvent event) {
...
+          // checkpoint boundary
           StatisticsEvent statisticsEvent =
               StatisticsEvent.createAggregatedStatisticsEvent(
-                  checkpointId, globalStatistics, aggregatedStatisticsSerializer);
+                  statistics, aggregatedStatisticsSerializer, false);
           for (int i = 0; i < context.currentParallelism(); ++i) {
+            // Ignore future return value for potential error, e.g. subtask down
             subtaskGateways.getSubtaskGateway(i).sendEvent(statisticsEvent);
           }
-
```
Comment (pvary) [inline flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsCoordinator.java:208]: Nit: needed because of a block?

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsOperator.java [hunk 7]
```diff
@@ -103,14 +105,41 @@ public void initializeState(StateInitializationContext context) throws Exception
...
+                  + "not matching downstream parallelism due to rescale. "
+                  + "Old parallelism is {}, new parallelism is {}",
+              operatorName,
+              subtaskIndex,
+              restoredStatistics.keySamples().length + 1,
+              downstreamParallelism);
+          // Asynchronously request the latest global statistics calculated with new downstream
+          // parallelism. It is possible events may have started flowing before coordinator responds
+          // with global statistics. In this case, range partitioner would just blindly shuffle
+          // records in round robin fashion.
```
Comment (pvary) [inline flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsOperator.java:134]: Maybe we could be better off if we use the old ranges with some heuristics?
- If we have higher number of subtasks, just use the old ranges, and leave idle tasks
- If we have lower number of subtasks, then use modulo?

Or the communication is fast enough, that it doesn't worth the complexity?

File: flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/sink/shuffle/TestAggregatedStatisticsTracker.java [hunk 8]
```diff
@@ -246,7 +254,14 @@ public void receiveCompletedStatisticsEvent(StatisticsType type) {
...
                   CHAR_KEYS.get("b"), 3L));
     } else {
-      assertThat(completedStatistics.rangeBounds()).containsExactly(CHAR_KEYS.get("a"));
+      assertThat(completedStatistics.keySamples())
+          .containsExactly(
+              CHAR_KEYS.get("a"),
+              CHAR_KEYS.get("b"),
+              CHAR_KEYS.get("b"),
+              CHAR_KEYS.get("a"),
+              CHAR_KEYS.get("a"),
```
Comment (pvary) [inline flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/sink/shuffle/TestAggregatedStatisticsTracker.java:263]: So the key samples are not ordered, just randomly selected samples from the incoming records?

Comment (pvary) [reply flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/sink/shuffle/TestAggregatedStatisticsTracker.java:263]: I have expected a bit more (based on my tests with sketches over long values), but having arbitrary keys probably doesn't allow better approximation

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/AggregatedStatistics.java (hunk 1 above)
Comment (stevenzwu) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/AggregatedStatistics.java:38]: that is definitely a fair question. I also thought about it. I can make the change if you think it is better.

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsOperator.java (hunk 5 above)
Comment (stevenzwu) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsOperator.java:146]: yeah. because Iceberg sink flush and commit at checkpoint boundary, switching at checkpoint boundary allows all subtasks to switch to the new stats for the same checkpoint cycle. Otherwise, some records are shuffled based on old stats and some are shuffled based on new stats.

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsOperator.java (hunk 7 above)
Comment (stevenzwu) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsOperator.java:134]: great question. I also thought about the fallback heuristic. Initially, I was thinking maybe start with simple and we can improve this part if it turns out to be a problem. communication should be relatively fast (maybe a few to 10 ms) if parallelism/fan-out is not very high.

> If we have higher number of subtasks, just use the old ranges, and leave idle tasks

scale-up doesn't require any code change. it works like this already

> If we have lower number of subtasks, then use modulo?

I agree modulo could be a sensible strategy
* pro: better clustering than round-robin
* con: uneven distribution. some subtasks may get double the loads than the other subtasks. but if the stats refresh is fast (like less than a dozen of ms). maybe this is not a concern.

Hence, I am in favor of implementing the fallback behavior for rescale

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsCoordinator.java (hunk 6 above)
Comment (stevenzwu) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsCoordinator.java:208]: haha. this is a good question. I guess we were just avoiding the complexity of error handling for completable future. 

Now coming back to the error handling, what we really need is a reconciliation process/protocol. I think we can compute a hash code / signature in the `GlobalStatistics`. Subtasks always send the `RequestStatisticsEvent` with the signature of its local copy upon startup. Coordinator can check if the signature matches or not. if not, resend the latest statistics.

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsOperator.java (hunk 7 above)
Comment (stevenzwu) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsOperator.java:134]: BTW, we haven't added the SketchRangePartitioner yet. fallback handling would be implemented there. so it will be out of the scope for this PR.

File: flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/sink/shuffle/TestAggregatedStatisticsTracker.java (hunk 8 above)
Comment (stevenzwu) [reply flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/sink/shuffle/TestAggregatedStatisticsTracker.java:263]: Sketch returns the array that captures the reservoir samples. before array is full, samples were added to the array in order.

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/AggregatedStatistics.java (hunk 1 above)
Comment (pvary) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/AggregatedStatistics.java:38]: Let's talk about this offline. I don't see all the cons and pros ATM.

Comment (pvary) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/AggregatedStatistics.java:38]: Reading through the code again, I'm more-and-more convinced, that we have 2 different objects here:
- RangeBounds (former global statistics) - key-values of the weights used by the partitioner with hash
- Statistics (former completed statistics) - Sketch or Map without hash, but full of data

I think we are just confusing them because of historical reasons.

**Comment by pvary** [comment]
@stevenzwu: Unrelated question, but come up when I have reading the PR:
- What happens in CDC cases, where the partition of the record might end up on multiple ranges. Did we make sure that the records with the same id go to the same subtask, so we can create a positional delete if there are multiple changes for the same id?

**Comment by stevenzwu** [comment]
> * What happens in CDC cases, where the partition of the record might end up on multiple ranges. Did we make sure that the records with the same id go to the same subtask, so we can create a positional delete if there are multiple changes for the same id?

CDC/Upsert should use existing hash distribution in `FlinkSink`. range distribution shouldn't be used in this case

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/AggregatedStatistics.java (hunk 1 above)
Comment (stevenzwu) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/AggregatedStatistics.java:38]: Yes, conceptually we have two types of objects here. for Map statistics, there is no difference btw global statistics and completed statistics, as there is no further reduction in stats size. For sketch statistics, global statistics is a lot smaller with range bounds.

We can introduce two types `CompleteStatistics` and `GlobalStatistics`. We can also introduce a base type `AggregatedStatistics`. I am trying to avoid duplicate the `AggregatedStatisticsSerializer` as it can work for both types. Maybe generics can enable code reused and solve the most duplications.

**Comment by pvary** [comment]
> > * What happens in CDC cases, where the partition of the record might end up on multiple ranges. Did we make sure that the records with the same id go to the same subtask, so we can create a positional delete if there are multiple changes for the same id?
> 
> CDC/Upsert should use existing hash distribution in `FlinkSink`. range distribution shouldn't be used in this case

I think we can overlay the hash distribution above the ranges, and we could make it work, but I undestand your reluctance to try to grab too much in one go.

**Comment by stevenzwu** [comment]
> I think we can overlay the hash distribution above the ranges,

Not sure if we want that, hash distribution (keyBy) is simple and low overhead. Range distribution requires statistics collection.

**Comment by pvary** [comment]
> > I think we can overlay the hash distribution above the ranges,
> 
> Not sure if we want that, hash distribution (keyBy) is simple and low overhead. Range distribution requires statistics collection.

If you partition your orders by time, and need to update the order if it was canceled, the your  key/partition is not equally distributed, and hashing is probably not a good option.

I would like to see the range partitioning as a precursor for writing ordered files with Flink. If we use similar constructs as the Flink SQL ORDER_BY then we can order the rows before writing them out. If we want to do this for CDC streams then we need to send the records with the same id to the same subtask.

Again, not something immediate, but might worth revisiting later.

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/AggregatedStatistics.java (hunk 1 above)
Comment (stevenzwu) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/AggregatedStatistics.java]: last commit separated out the two types

**Comment by stevenzwu** [comment]
> > > I think we can overlay the hash distribution above the ranges,
> > 
> > 
> > Not sure if we want that, hash distribution (keyBy) is simple and low overhead. Range distribution requires statistics collection.
> 
> If you partition your orders by time, and need to update the order if it was canceled, the your key/partition is not equally distributed, and hashing is probably not a good option.
> 
> I would like to see the range partitioning as a precursor for writing ordered files with Flink. If we use similar constructs as the Flink SQL ORDER_BY then we can order the rows before writing them out. If we want to do this for CDC streams then we need to send the records with the same id to the same subtask.
> 
> Again, not something immediate, but might worth revisiting later.

I am happy to discuss it here. Agree that it is not sth we want to address in this PR.

ORDER_BY is essentially the SortOrder defined in table properties. Note that currently Flink writer doesn't sort rows within a data file. Range partitioner only range split keys across files for better clustering.

In the example of orders table partitioned by time (say hourly), the primary keys would be `(hour(ts), order_id)` tuple. hash distribution (keyBy) can work and ensure correctness. You are saying range distribution would work better, because the better clustering of `order_id` within an hourly partition, correct? With hash distribution, each subtask writes the number of data files equals to the number of hours in a checkpoint cycle. Range distribution can handles the event time skew (recent hours have more data than distant past). Yes, I agree with that assessment. There is some challenges with handling new values (like new hours as time advances). Right now, it is handled by round-robin, as there is no assumption of `SortOrder` should be handled as primary keys. Primary keys are better/safer handled as hash distribution.

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/GlobalStatistics.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+/**
+ * This is the globally aggregated statistics by the coordinator. It is what is sent to the operator
+ * subtasks for range partitioning.
+ */
+class GlobalStatistics extends AggregatedStatistics {
+  GlobalStatistics(
+      long checkpointId,
+      StatisticsType type,
+      Map<SortKey, Long> keyFrequency,
+      SortKey[] rangeBounds) {
```
Comment (pvary) [inline flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/GlobalStatistics.java:33]: Don't we just send the `rangeBounds` back as a `GlobalStatistics`?

Comment (stevenzwu) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/GlobalStatistics.java:33]: yes, we do. `rangeBounds` is also an array of `SortKey`. 

Here is an example from `SketchUtil`
```
   * To understand how range bounds are used in range partitioning, here is an example for human
   * ages with 4 partitions: [15, 32, 60]. The 4 ranges would be
   *
   * <ul>
   *   <li>age <= 15
   *   <li>age > 15 && age <= 32
   *   <li>age >32 && age <= 60
   *   <li>age > 60
   * </ul>
```

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsOperator.java (hunk 7 above)
Comment (stevenzwu) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsOperator.java]: added the SketchRangePartitioner and RangePartitioner to this PR. so the scope is bigger now

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/AggregatedStatistics.java (hunk 1 above)
Comment (stevenzwu) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/AggregatedStatistics.java:35]: I have removed the `AggregatedStatistics` base class. With the change to the `GlobalStatistics` on Map key assignment, the base class doesn't make much sense anymore. Now `GlobalStatistics` is used by partitioner and `CompletedStatistics` are raw aggregated stats.

File: flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/RangePartitioner.java [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+  private Partitioner<RowData> delegatePartitioner(GlobalStatistics statistics) {
+    if (statistics.type() == StatisticsType.Map) {
+      return new MapRangePartitioner(schema, sortOrder, statistics.mapAssignment());
+    } else if (statistics.type() == StatisticsType.Sketch) {
+      return new SketchRangePartitioner(schema, sortOrder, statistics.rangeBounds());
+    } else {
+      throw new IllegalArgumentException(
+          String.format("Invalid statistics type: %s. Should be Map or Sketch", statistics.type()));
+    }
+  }
```
Comment (pvary) [inline flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/RangePartitioner.java:81]: I still struggling a bit creating a good mental model for the `GlobalStatistics` distribution.

I feel that separating out the `GlobalStatistics` would be good if we could create a data structure which could be used by both `MapRangePartitioner` and `SketchRangePartitioner`, but sadly these partitioners require different input data.

Would it make sense to serialize/parametrize the partitioner on the JM side and send it in the records instead of the `GlobalStatistics` in the `StatisticsOrRecord`?
Like `PartitionerOrRecord`, and on the partitioner side we just call the `partition` and forget it.

Would this be easier to understand for others?

Comment (stevenzwu) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/RangePartitioner.java:81]: > I feel that separating out the GlobalStatistics would be good if we could create a data structure which could be used by both MapRangePartitioner and SketchRangePartitioner

`GlobalStatistics` is for that purpose. it contains either map assignment or range bounds.

> Would it make sense to serialize/parametrize the partitioner on the JM side and send it in the records instead of the GlobalStatistics in the StatisticsOrRecord?

not sure if there is much benefit to serialize the delegate partitioner on the JM side and ship to the subtasks. that would require Java serialization and make checkpoint state more complex for schema evolution.

Comment (pvary) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/RangePartitioner.java:81]: My issue with the GlobalStatistics is that it is not really global. The main payload (mapAssignments and rangeBounds) are different for the different Partitioners. So basically we pretend that we send a statistics, but in reality the data already describes the partitioner. Maybe it would be cleaner to accept this, and mirror this change in the code/class names etc.

Instead of writing a serializer for the GlobalStatistics, we can write a serializer for the respective partitioners. I don't see this as a blocker.

I don't have a very strong opinion on this, and I wanted you to understand why I feel that the code is a bit awkward in this case. What is usually a yellow flag for me, is to have an object where some fields are mutually exclusive. I try to examine those again to understand if we have a single object, or we just merged 2 objects to a single one.

Comment (stevenzwu) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/RangePartitioner.java:81]: `GlobalStatistics` is global in the sense it is aggregated statistics by coordinator. We have a facade/proxy partitioner that would delegate the partition decision to underline partitioner based on the statistics type.

Comment (pvary) [reply flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/RangePartitioner.java:81]: After our offline discussion I understand your points better. I understand that the concept is
- That we collect the statistics
- Then use the statistics to create a partitioner
and we try to stick to the concept.

I still think it is confusing, that currently we collect 2 types of statistics and each of them are tightly coupled to the 2 types of partitioners we have. (`Type.Map` is always used by `MapRangePartitioner`, and `Type.Sketch` is always used by `SketchRangePartitioner`). We could create the `MapRangePartitioner` and the `SketchRangePartitioner` on the coordinator side, and send them to a PartitionerExecutor which just deserializes them and runs them. So the real logic would be in a single place (`DataStatisticsCoordinator`).

That said, the improvement is just coding style preference, as the Partitioners still need to serialize the underlying statistics, so the performance would be the same. 

So we can move forward with your proposed solution.
Thanks for the discussion!

**Review by pvary — APPROVED** [review]
Hi Steven,
I think we discussed every comment.
Could we run the tests one more time before merging? It was a long time ago when they were running, and it might be good to double check before merging.

**Comment by stevenzwu** [comment]
> Hi Steven, I think we discussed every comment. Could we run the tests one more time before merging? It was a long time ago when they were running, and it might be good to double check before merging.

let me rebase with the latest main branch

**Comment by stevenzwu** [comment]
thanks @pvary for the review
